onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc

```diff
@@ -0,0 +1,476 @@
+  size_t sizeAll = 0;
+
+  for (int i = 0; i < initializer_size; ++i) {
+    const auto& tensor = model_proto_.graph().initializer(i);
+    const auto& name = tensor.name();
+    if (Contains(skipped_initializers_, name))
+      continue;
+
+    Shape shape;
+    for (auto dim : tensor.dims()) {
+      shape.push_back(SafeInt<uint32_t>(dim));
+    }
+
+    Type type = Type::TENSOR_FLOAT32;
+    switch (tensor.data_type()) {
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT:
+        type = Type::TENSOR_FLOAT32;
+        break;
+      default:
+        // TODO: support other type
+        ORT_THROW("The initializer of graph doesn't have valid type, name: " +
+                  name + " type: " + std::to_string(tensor.data_type()));
+        break;
+    }
+
+    OperandType operand_type(type, shape);
+    shaper_.AddShape(name, operand_type.dimensions);
+
+    auto index = AddNewOperand(name, operand_type);
+    const size_t size = operand_type.GetOperandBlobByteSize();
+    const size_t paddedSize = GetPaddedByteSize(size);
+    sizeAll += paddedSize;
+    initializers[i] = std::make_tuple(index, size, paddedSize);
+  }
+
+  // 2nd pass copies all the initializer data into NNAPI shared memory
+  nnapi_model_->mem_initializers_ =
+      std::make_unique<Model::NNMemory>(nnapi_, "mem_initializers_", sizeAll);
+
+  // 2nd pass to copy all the initializers into shared memory
+  size_t offset = 0;
+  for (int i = 0; i < initializer_size; ++i) {
+    const auto& tensor = model_proto_.graph().initializer(i);
+    if (Contains(skipped_initializers_, tensor.name()))
+      continue;
+
+    Index index;
+    size_t size, paddedSize;
+    std::tie(index, size, paddedSize) = initializers[i];
+    const char* src = nullptr;
+    if (tensor.data_type() == ONNX_NAMESPACE::TensorProto_DataType_FLOAT) {
+      src = tensor.float_data().empty()
+                ? tensor.raw_data().data()
+                : reinterpret_cast<const char*>(tensor.float_data().data());
+    } else {
+      ORT_THROW("The initializer of graph doesn't have valid type, name: " +
+                tensor.name() + " type: " + std::to_string(tensor.data_type()));
+    }
+
+    uint8_t* dest = nnapi_model_->mem_initializers_->get_data_ptr() + offset;
+    memcpy(dest, src, size);
+    SetOperandValue(index, nnapi_model_->mem_initializers_.get(), size, offset);
+    offset += paddedSize;
+  }
+}
+
+void ModelBuilder::RegisterModelInputs() {
+  for (int32_t input_idx = 0; input_idx < model_proto_.graph().input_size(); input_idx++) {
+    const auto& input(model_proto_.graph().input(input_idx));
+    std::string input_name = input.name();
+
+    {  // input should not be an initializer
+      if (Contains(operands_, input_name))
+        continue;
+
+      if (Contains(initializers_, input_name))
+        continue;
+    }
+
+    Shaper::Shape shape;
+    for (const auto& dim : input.type().tensor_type().shape().dim()) {
+      shape.push_back(SafeInt<uint32_t>(dim.dim_value()));
+    }
+
```

Comment:
Does this need to handle where there's a dim_param or neither? If not, a comment saying why would help explain why it's safe to assume dim_value is present.

Reply:
Added comments